File: calcpp/Functions/Operators.py
# -*- coding: utf-8 -*-
import yaml

try:
    from yaml import CLoader as Loader
except ImportError:
    from yaml import Loader


from sourcegen import SourceGenerator

with open("../symbols.yml") as file:
    symbols = yaml.load(file, Loader)


if __name__ == "__main__":
    operators = symbols["operators"]

    opinfo = [
        (
            operator,
            (
                (i + 1)
                * (1 if metadata is not None and "rightAsociative" in metadata else -1),
                (1 if metadata is not None and "singleOperator" in metadata else 0),
            ),
        )
        for i, group in enumerate(operators)
        for operator, metadata in group.items()
    ]
    opinfo.sort()

    with SourceGenerator("Operators.cc") as source:
        source["size"].set_text(len(operators))
        source["offset"].set_text(len(symbols["tokens"]))

        source["precedences"].set_text(
            (str(p) for op, (p, s) in opinfo), delimiter=", ", wrap_width=80
        )
        source["single"].set_text(
            (str(s) for op, (p, s) in opinfo), delimiter=", ", wrap_width=80
        )

File: calcpp/Functions/Operators.cc

#include "Operators.h"
#include "Functions.h"

namespace calcpp {

    namespace operators {

        constexpr const int size =
            // begin sourcegen size
            17
            // end sourcegen
            ;
        constexpr const int offset =
            // begin sourcegen offset
            20
            // end sourcegen
            ;
        constexpr const int precedences[size] = {
            // begin sourcegen precedences
            -8, -5, -11, -15, -16, -2, -7, -17, -17, -2, -1, -3, -3, -3, -3, -2,
            -9, -15, -9, -15, -10, -10, -12, -2, -10, -10, -12, -2, -2, -14, -2,
            -15, -2, -6, -4, -13, -2, -17, -2, -17, -17, -15, -2, -15, -15, -2,
            -16, -10, -2
            // end sourcegen
        };
        constexpr const bool single[size] = {
            // begin sourcegen single
            0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0,
            0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
            0, 0, 0
            // end sourcegen
        };

        inline int _abs(int x) { return x >= 0 ? x : -x; }

        int precedence(const Token::Kind kind) {
            return _abs(precedences[int(kind) - offset]);
        }
        bool rightAssociative(const Token::Kind kind) {
            return precedences[int(kind) - offset] > 0;
        }

        bool singleOperator(const Token::Kind kind) {
            return single[int(kind) - offset];
        }

    }  // namespace operators

    constexpr bool isOperator(const Token::Kind kind) {
        return (int(kind) - operators::offset) >= 0;
    }

    expression getOperator(const Token::Kind kind) { return None; }

}  // namespace calcpp
